source/extensions/filters/http/common/aws/signer_impl.cc

```diff
@@ -0,0 +1,256 @@
+const Http::LowerCaseString SignerImpl::X_AMZ_CONTENT_SHA256{"x-amz-content-sha256"};
+
+void SignerImpl::sign(Http::Message& message) const {
+  const auto& credentials = credentials_provider_->getCredentials();
+  if (!credentials.accessKeyId() || !credentials.secretAccessKey()) {
+    return;
+  }
+  const auto& region = region_provider_->getRegion();
+  if (!region) {
+    throw EnvoyException("Could not determine AWS region");
+  }
+  auto& headers = message.headers();
+  if (credentials.sessionToken()) {
+    headers.addCopy(X_AMZ_SECURITY_TOKEN, credentials.sessionToken().value());
+  }
+  const auto long_date = LONG_DATE_FORMATTER.now(time_source_);
+  const auto short_date = SHORT_DATE_FORMATTER.now(time_source_);
+  headers.addCopy(X_AMZ_DATE, long_date);
+  const auto content_hash = createContentHash(message);
+  headers.addCopy(X_AMZ_CONTENT_SHA256, content_hash);
+  // Phase 1: Create a canonical request
+  const auto canonical_headers = canonicalizeHeaders(headers);
+  const auto signing_headers = createSigningHeaders(canonical_headers);
+  const auto canonical_request =
+      createCanonicalRequest(message, canonical_headers, signing_headers, content_hash);
+  ENVOY_LOG(debug, "Canonical request:\n{}", canonical_request);
+  // Phase 2: Create a string to sign
+  const auto credential_scope = createCredentialScope(short_date, region.value());
+  const auto string_to_sign = createStringToSign(canonical_request, long_date, credential_scope);
+  ENVOY_LOG(debug, "String to sign:\n{}", string_to_sign);
+  // Phase 3: Create a signature
+  const auto signature = createSignature(credentials.secretAccessKey().value(), short_date,
+                                         region.value(), string_to_sign);
+  // Phase 4: Sign request
+  const auto authorization_header = createAuthorizationHeader(
+      credentials.accessKeyId().value(), credential_scope, signing_headers, signature);
+  ENVOY_LOG(debug, "Signing request with: {}", authorization_header);
+  headers.addCopy(Http::Headers::get().Authorization, authorization_header);
+}
+
+std::string SignerImpl::createContentHash(Http::Message& message) const {
+  if (!message.body()) {
+    return HASHED_EMPTY_STRING;
+  }
+  return Hex::encode(hash(*message.body()));
+}
+
+std::string SignerImpl::createCanonicalRequest(
+    Http::Message& message, const std::map<std::string, std::string>& canonical_headers,
+    const std::string& signing_headers, const std::string& content_hash) const {
+  const auto& headers = message.headers();
+  std::stringstream out;
+  // Http method
+  const auto* method_header = headers.Method();
+  if (method_header == nullptr || method_header->value().empty()) {
+    throw EnvoyException("Message is missing :method header");
+  }
+  out << method_header->value().c_str() << "\n";
+  // Path
+  const auto* path_header = headers.Path();
+  if (path_header == nullptr || path_header->value().empty()) {
+    throw EnvoyException("Message is missing :path header");
```

Comment:
I don't think either the method or path should be missing by construction, so you can probably omit these.